### PR TITLE
Add RFC for more Iter methods

### PR DESCRIPTION
**Pull request by Theodus**
[Rendered](https://github.com/Theodus/rfcs/blob/iter/text/0000-more-iter-methods.md)

### CODE REVIEW

**Comment by SeanTAllen** [comment]
Does the "how do we test this" translate to "unit tests"?

## 

For "how do we teach this", I think that the high level package documentation in itertools.pony needs to be updated for previous iter changes and this. that package level documentation looks pretty out of date. agreed?

**Comment by Theodus** [comment]
@SeanTAllen Agreed. I'll make changes to update the documentation and clarify that unit tests are required.

File: text/0000-more-iter-methods.md [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+The purpose of these additional methods is to increase the functionality of the Iter class.
+
+# Detailed design
+
+The following methods will be added to the Iter class:
+```pony
+fun ref all(f: {(A!): Bool ?} box): Bool =>
+    """
+    Return false if at least one value of the iterator fails to match the
+    predicate `f`, otherwise return true.
```
Comment (jemc) [inline text/0000-more-iter-methods.md]: Is it short-circuiting?  (Do all of elements get the function executed, or does execution stop when a `false` is reached?)  This should be made clear in the doc string.

File: text/0000-more-iter-methods.md [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    """
+
+fun ref any(f: {(A!): Bool ?} box): Bool =>
+    """
+    Return true if at least one value of the iterator matches the predicate `f`,
+    otherwise return false.
```
Comment (jemc) [inline text/0000-more-iter-methods.md]: Is it short-circuiting?  (Do all of elements get the function executed, or does execution stop when a `true` is reached?)  This should be made clear in the doc string.

**Comment by SeanTAllen** [comment]
@Theodus you ready for final comment on this?

**Comment by Theodus** [comment]
@SeanTAllen Yes I am.

**Comment by Theodus** [comment]
The method `run` has been added to iterate over the values of the iterator without a for loop.

**Comment by SeanTAllen** [comment]
@Theodus `run` is missing a return type. or if it isn't, I don't understand how it is used.

Lots of examples in the package docstring would be a good addition.

**Comment by Theodus** [comment]
@SeanTAllen would there be a reason to return a spent iterator?

**Comment by SeanTAllen** [comment]
@theodus how would you use this run method? I don't understand how it's useful. I must be missing something.

**Comment by Theodus** [comment]
@SeanTAllen With the `run` method you can do the following:

The following snippet prints the numbers 1-3.

``` pony
let xs = Iter[I64](Range[I64](1, 4))
  .map[None](lambda(x: I64)(env) => env.out.print(x.string()) end)
  .run()
```

Without the `run` method, you would have to do the following:

``` pony
let xs = Iter[I64](Range[I64](1, 4))
  .map[None](lambda(x: I64)(env) => env.out.print(x.string()) end)

for x in xs do None end
```

**Comment by SeanTAllen** [comment]
so map is lazy?

**Comment by Theodus** [comment]
@SeanTAllen Many of the methods return iterators rather than evaluating, so yes.

**Comment by SeanTAllen** [comment]
Ah, that makes sense. Definitely something good for the documentation. Discuss which are "lazy" vs which are "eager".

**Pull request merged**